presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java

```diff
@@ -0,0 +1,229 @@
+import io.prestosql.sql.planner.plan.PlanNodeId;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+import static com.google.common.base.Verify.verify;
+import static io.prestosql.SystemSessionProperties.getDynamicFilteringMaxPerDriverByteSize;
+import static io.prestosql.SystemSessionProperties.getDynamicFilteringMaxPerDriverRowCount;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This operator acts as a simple "pass-through" pipe, while saving its input pages.
+ * The collected pages' value are used for creating a run-time filtering constraint (for probe-side table scan in an inner join).
+ * We support only small build-side pages (which should be the case when using "broadcast" join).
+ */
+public class DynamicFilterCollectOperator
+        implements Operator
+{
+    public static class DynamicFilterCollectOperatorFactory
+            implements OperatorFactory, Consumer<Optional<TupleDomain<ColumnHandle>>>
+    {
+        private final int operatorId;
+        private final PlanNodeId planNodeId;
+        private final DynamicFilter dynamicFilter;
+
+        private TupleDomain<ColumnHandle>[] partitions;
+        private int partitionCount;
+        private final Object partitionLock;
+
+        public DynamicFilterCollectOperatorFactory(int operatorId, PlanNodeId planNodeId, DynamicFilter dynamicFilter, int partitionCount)
+        {
+            this.operatorId = operatorId;
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.dynamicFilter = requireNonNull(dynamicFilter);
+
+            this.partitions = new TupleDomain[partitionCount];
+            this.partitionLock = new Object();
+        }
+
+        @Override
+        public void accept(Optional<TupleDomain<ColumnHandle>> dynamicPredicate)
+        {
+            // Called by each DynamicFilterCollectOperator instance (when collection is over).
+            synchronized (partitionLock) {
+                verify(this.partitionCount < partitions.length);
+                if (partitions == null) {
+                    return;
+                }
+                if (!dynamicPredicate.isPresent()) {
+                    partitions = null;  // drop all collected predicates if one of partitions' results is too large.
+                    return;
+                }
+                partitions[this.partitionCount++] = dynamicPredicate.get();
+                if (this.partitionCount == partitions.length) {
+                    TupleDomain<ColumnHandle> dynamicConstraint = Arrays.stream(partitions).reduce(TupleDomain.none(), TupleDomain::columnWiseUnion);
+                    // NOTE: may result in a bit more relaxed constraint if there are multiple columns and multiple rows.
+                    // See the comment at TupleDomain::columnWiseUnion() for more details.
+                    dynamicFilter.accept(dynamicConstraint);
```

Comment:
callback should not be called within critical section, see: `io.prestosql.operator.exchange.LocalExchangeSource#addPage`

Reply:
Should be fixed by resolving https://github.com/prestosql/presto/pull/931#discussion_r302104425 at c343902c15.